### PR TITLE
SCP-259 Production of a CostModel

### PR DESCRIPTION
**Pull request by reactormonk**
Could improve:
- complete CSV, in production, coming soon™
- clean up the models in `ExBudgeting.hs` with some generic magic.

### CODE REVIEW

File: language-plutus-core/budgeting-bench/Bench.hs [hunk 1]
```diff
@@ -1,5 +1,158 @@
...
+benchBytestringOperations :: BuiltinName -> Benchmark -- TODO the numbers are a bit too big here
+benchBytestringOperations name = createTwoTermBuiltinBench @Integer @BSL.ByteString name numbers bsA
+    where
+        bsA = (expToBenchingBytestring seedA . fromInteger) <$> expsToBench
+        numbers = expToBenchingInteger <$> expsToBench
+
+createTwoTermBuiltinBench :: (DefaultUni `Includes` a, DefaultUni `Includes` b) => BuiltinName -> [(a, ExMemory)] -> [(b, ExMemory)] -> Benchmark
+createTwoTermBuiltinBench name as bs =
+    bgroup (show name) $
+        (flip fmap) as (\(x, xMem) ->
```
Comment (effectfully) [inline language-plutus-core/budgeting-bench/Bench.hs:68]: I think `flip fmap as f` is better spelled as `as <&> f`.

And in other places.

File: language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExMemory.hs [hunk 2]
```diff
@@ -119,7 +123,7 @@ instance ExMemoryUsage () where
   memoryUsage _ = 0 -- TODO or 1?
 
 instance ExMemoryUsage Integer where
-  memoryUsage _ = 2 -- TODO
+  memoryUsage i = ExMemory (smallInteger (integerLog2# (abs i) `quotInt#` (integerToInt 60))) -- assume 60bit size
```
[the remark this answers is not in the document]
Comment (reactormonk) [reply language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExMemory.hs:126]: Where did you find said `integerLog2`?

File: language-plutus-core/budgeting-bench/Bench.hs [hunk 3]
```diff
@@ -1,5 +1,158 @@
...
+
+benchSameTwoByteStrings :: BuiltinName -> Benchmark
+benchSameTwoByteStrings name = createTwoTermBuiltinBench name bsA bsA
+    where
+        bsA = (expToBenchingBytestring seedA . fromInteger) <$> expsToBench
+
+benchTwoByteStrings :: BuiltinName -> Benchmark
+benchTwoByteStrings name = createTwoTermBuiltinBench name bsA bsB
+    where
+        bsA = (expToBenchingBytestring seedA . fromInteger) <$> expsToBench
```
[the remark this answers is not in the document]
Comment (reactormonk) [reply language-plutus-core/budgeting-bench/Bench.hs]: That'd explain why my benchies kept crashing.

File: language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExMemory.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (effectfully) [reply language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExMemory.hs]: I was looking [here](https://hackage.haskell.org/package/arithmoi-0.4.0.3/docs/src/Math-NumberTheory-Logarithms.html#integerLog2).

[the remark this answers is not in the document]
Comment (reactormonk) [reply language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExMemory.hs]: If you look at the implementation, it does exactly the same :-)

```
integerLog2 :: Integer -> Int
integerLog2 n
  | n < 1       = error "Math.NumberTheory.Logarithms.integerLog2: argument must be positive"
  | otherwise   = I# (integerLog2# n)
```

File: language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExBudgeting.hs [hunk 4]
```diff
@@ -197,48 +202,99 @@ data CostModel =
...
 
 runCostingFunOneArgument :: CostingFunOneArgument -> ExMemory -> ExBudget
 runCostingFunOneArgument
-    (OneArgumentConstantCost c) _ = ExBudget (ExCPU c) (ExMemory c)
-runCostingFunOneArgument
-    (OneArgumentLogLinearCost factor) (ExMemory size1) = ExBudget (ExCPU $ size1 * factor) (ExMemory $ size1 * factor)
+    (CostingFunOneArgument cpu mem) mem1 =
+        ExBudget (ExCPU (runOneArgumentModel cpu mem1)) (ExMemory (runOneArgumentModel mem mem1))
+
+runOneArgumentModel :: ModelOneArgument -> ExMemory -> Integer
```
Comment (michaelpj) [inline language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExBudgeting.hs:230]: Have we got a note somewhere about the general approach with these `Model` types?

File: language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExBudgetingDefaults.hs [hunk 5]
```diff
@@ -1,9 +1,14 @@
 {-# LANGUAGE TemplateHaskell #-}
 module Language.PlutusCore.Evaluation.Machine.ExBudgetingDefaults where
 
-import           Data.Aeson.THReader
 import           Language.PlutusCore.Evaluation.Machine.ExBudgeting
 
+-- import           Data.Aeson.THReader
```
Comment (michaelpj) [inline language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExBudgetingDefaults.hs]: Commented out code

File: nix/haskell.nix [hunk 6]
```diff
@@ -73,6 +73,9 @@ let
             # plc-agda is compiled from the Haskell source files generated from the Agda
             plc-agda.src = "${metatheory.plutus-metatheory-compiled}/share/agda";
 
+            language-plutus-core.components.benchmarks.language-plutus-core-create-cost-model.build-tools =
+              lib.mkForce [(pkgs.rWrapper.override { packages = with pkgs.rPackages; [tidyverse dplyr stringr MASS]; } )];
```
Comment (michaelpj) [inline nix/haskell.nix:77]: Hm, I feel like we should be able to do this in a nicer way but looks fine for now.

File: shell.nix [hunk 7]
```diff
@@ -29,6 +29,8 @@ with packageSet; haskell.packages.shellFor {
     dev.packages.purty
     dev.packages.purs
     dev.packages.spago
+    # From haskell.nix
+    (pkgs.rWrapper.override { packages = with pkgs.rPackages; [tidyverse dplyr stringr MASS]; } )
```
Comment (michaelpj) [inline shell.nix:33]: Hm, I'd have hoped this got pulled in as a build dep by the shell :thinking:

File: language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExMemory.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (effectfully) [reply language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExMemory.hs]: I did look at the implementation, which is why I suggested using some other function than `integerLog2#`, which seems to rely on the `n < 1` invariant, which it doesn't seem to check. You could introduce your own wrapper around `integerLog2#` doing the necessary invariant check.

File: language-plutus-core/test/Evaluation/Machines/Budget/Fib/3.plc.golden [hunk 8]
```diff
@@ -2,7 +2,7 @@
 , ({ tally: ({ BBuiltin (StaticStagedBuiltinName SubtractInteger) causes ({ cpu: 4
 | mem: 4
 })
-| BBuiltin (StaticStagedBuiltinName AddInteger) causes ({ cpu: 2
+| BBuiltin (StaticStagedBuiltinName AddInteger) causes ({ cpu: 0
 | mem: 2
```
Comment (effectfully) [inline language-plutus-core/test/Evaluation/Machines/Budget/Fib/3.plc.golden:6]: Was `cpu: 2, mem: 2` and now `cpu: 0, mem: 2`?

File: language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExMemory.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (reactormonk) [reply language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExMemory.hs]: Yeah, I added said check inline. Dunno if that's the right way to go.

File: language-plutus-core/create-cost-model/CreateModel.hs [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    Right x  -> pure x
+
+
+addInteger :: MonadR m => (SomeSEXP (Region m)) -> m CostingFunTwoArguments
+addInteger df = do
+  addInt <- filterDF "AddInteger" df
+  cpuModel <- linearAdditiveModel addInt
+  pure $ CostingFunTwoArguments (ModelTwoArgumentsAddedSizes cpuModel) def
+
+subtractInteger :: MonadR m => (SomeSEXP (Region m)) -> m CostingFunTwoArguments
```
[the remark this answers is not in the document]
Comment (reactormonk) [reply language-plutus-core/create-cost-model/CreateModel.hs]: Yeah, I don't have any models for them just yet, so they'll be filled in later.

File: language-plutus-core/language-plutus-core.cabal [hunk 10]
```diff
@@ -314,4 +316,32 @@ benchmark language-plutus-core-budgeting-bench
         deepseq -any,
         lens -any,
         directory -any,
-        integer-gmp -any
+        integer-gmp -any,
+        hedgehog -any
+
+benchmark language-plutus-core-create-cost-model
```
[the remark this answers is not in the document]
Comment (reactormonk) [reply language-plutus-core/language-plutus-core.cabal:322]: I don't know really. I put it here because it's rather module-internal rather than something you want to use as an end-user, so I figured `executable` wouldn't work too well.

File: language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExBudgetingDefaults.hs (hunk 5 above)
Comment (reactormonk) [reply language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExBudgetingDefaults.hs]: That one's here for a reason :-)

File: nix/haskell.nix (hunk 6 above)
Comment (reactormonk) [reply nix/haskell.nix:77]: How would you do this nicer? I currently don't see another way.

File: shell.nix (hunk 7 above)
Comment (reactormonk) [reply shell.nix:33]: I _think_ it didn't work, I could try again.

File: language-plutus-core/test/Evaluation/Machines/Budget/Fib/3.plc.golden (hunk 8 above)
Comment (reactormonk) [reply language-plutus-core/test/Evaluation/Machines/Budget/Fib/3.plc.golden:6]: I never said the model was already calibrated properly :-)

File: language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExBudgetingDefaults.hs (hunk 5 above)
Comment (reactormonk) [reply language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExBudgetingDefaults.hs]: ... I mean I could make the TH nicer and emit a warning and a default when the model can't be loaded, but that does sound like some work. And I'm not even sure that's good, as if someone loads in the repo and gets a weird json for whatever reason, ignores the warning and ends up with an absolutely weird cost-model.

File: nix/haskell.nix (hunk 6 above)
Comment (michaelpj) [reply nix/haskell.nix:77]: I mean specifically the `mkForce`. Might need some changes in `haskell.nix`, but I'd have thought those should merge...

File: shell.nix (hunk 7 above)
Comment (reactormonk) [reply shell.nix:33]: Works.

File: language-plutus-core/create-cost-model/CreateModel.hs [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+
+
+addInteger :: MonadR m => (SomeSEXP (Region m)) -> m (CostingFun ModelTwoArguments)
+addInteger df = do
+  addInt <- filterDF "AddInteger" df
+  cpuModel <- linearAdditiveModel addInt
+  pure $ CostingFun (ModelTwoArgumentsAddedSizes cpuModel) def
+
+subtractInteger :: MonadR m => (SomeSEXP (Region m)) -> m (CostingFun ModelTwoArguments)
+subtractInteger _ = pure def
```
Comment (kwxm) [inline language-plutus-core/create-cost-model/CreateModel.hs:103]: Not sure what's going on here.  Are all of these waiting to be filled in with something similar to the addInteger case?

File: language-plutus-core/create-cost-model/CreateModel.hs [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+      paramConcatenate <- concatenate benchData
+      paramTakeByteString <- takeByteString benchData
+      paramDropByteString <- dropByteString benchData
+      paramSHA2 <- sHA2 benchData
+      paramSHA3 <- sHA3 benchData
+      paramVerifySignature <- verifySignature benchData
+      paramEqByteString <- eqByteString benchData
+      paramLtByteString <- ltByteString benchData
+      paramGtByteString <- gtByteString benchData
+      paramIfThenElse <- ifThenElse benchData
```
Comment (kwxm) [inline language-plutus-core/create-cost-model/CreateModel.hs:54]: This is kind of repetitive.  Could we use TH or something to simplify this, or would that be overkill?  (I don't have any particular opinion here, just wondering if that sort of thing's reasonable.)

File: language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExMemory.hs [hunk 13]
```diff
@@ -119,7 +123,7 @@ instance ExMemoryUsage () where
   memoryUsage _ = 0 -- TODO or 1?
 
 instance ExMemoryUsage Integer where
-  memoryUsage _ = 2 -- TODO
+  memoryUsage i = ExMemory (if i == 0 then 0 else smallInteger (integerLog2# (abs i) `quotInt#` (integerToInt 60))) -- assume 60bit size
```
Comment (kwxm) [inline language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExMemory.hs:126]: 60 bits is rather unexpected here.  What's going on?

File: marlowe-symbolic/src/Language/Marlowe/Analysis/FSSemantics.hs [hunk 14]
```diff
@@ -185,8 +185,8 @@ symEvalVal SlotIntervalStart symState = lowSlot symState
 symEvalVal SlotIntervalEnd symState = highSlot symState
 symEvalVal (UseValue valId) symState =
   M.findWithDefault (literal 0) valId (symBoundValues symState)
-symEvalVal (Cond cond v1 v2) symState = ite (symEvalObs cond symState) 
-                                            (symEvalVal v1 symState) 
+symEvalVal (Cond cond v1 v2) symState = ite (symEvalObs cond symState)
+                                            (symEvalVal v1 symState)
                                             (symEvalVal v2 symState)
 
 -- Symbolic version evalObservation
```
Comment (kwxm) [inline marlowe-symbolic/src/Language/Marlowe/Analysis/FSSemantics.hs:192]: Should this really be in here?  It's just a whitespace difference.  Was that done automatically?

File: plutus-ledger/src/Ledger/Scripts.hs [hunk 15]
```diff
@@ -55,7 +55,7 @@ module Ledger.Scripts(
 
 import qualified Prelude                              as Haskell
 
-import           Codec.Serialise                      (serialise, Serialise)
+import           Codec.Serialise                      (Serialise, serialise)
 import           Control.DeepSeq                      (NFData)
 import           Control.Monad.Except                 (MonadError, runExcept, throwError)
 import           Crypto.Hash                          (Digest, SHA256, hash)
```
Comment (kwxm) [inline plutus-ledger/src/Ledger/Scripts.hs:61]: Again, this has nothing to do with costing.  I presume it's happened because some formatting tool has been run over the entire codebase.  Do we have a policy about altering unrelated code in PRs?  I've done that myself a couple of times and it always made me feel uncomfortable.

File: language-plutus-core/create-cost-model/CreateModel.hs (hunk 11 above)
Comment (reactormonk) [reply language-plutus-core/create-cost-model/CreateModel.hs:103]: Exactly. I'll add in one or two more to make it clearer.

File: language-plutus-core/create-cost-model/CreateModel.hs (hunk 12 above)
Comment (reactormonk) [reply language-plutus-core/create-cost-model/CreateModel.hs:54]: If I had `sequenceRecord` from PureScript, I'd whip something up easily, but here I'm not sure if the benefits would make it worth it. Smaller code is cool when you expect heavy modification, but I see this one to be write-once mostly. So I'd vote for this version until I'll have to mess with it again, then I'll optimize it.

File: language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExMemory.hs (hunk 13 above)
Comment (reactormonk) [reply language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExMemory.hs:126]: I assume 60bits in an Integer, whereas you probably get 64bit per word. So it's a bit of an underestimation. Might even be +1 for all the bookkeeping information.

File: marlowe-symbolic/src/Language/Marlowe/Analysis/FSSemantics.hs (hunk 14 above)
Comment (reactormonk) [reply marlowe-symbolic/src/Language/Marlowe/Analysis/FSSemantics.hs:192]: Yup, I ran stylish.

File: plutus-ledger/src/Ledger/Scripts.hs (hunk 15 above)
Comment (reactormonk) [reply plutus-ledger/src/Ledger/Scripts.hs:61]: I run stylish, I commit.

Comment (michaelpj) [reply plutus-ledger/src/Ledger/Scripts.hs:61]: If you're running the same version of stylish as everyone else, then it shouldn't affect unrelated code, since that should already have been formatted :thinking: where do you get your `stylish-haskell` from?

File: language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExMemory.hs (hunk 13 above)
Comment (michaelpj) [reply language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExMemory.hs:126]: Sounds like it needs a comment :)

File: language-plutus-core/language-plutus-core.cabal (hunk 10 above)
[the remark this answers is not in the document]
Comment (michaelpj) [reply language-plutus-core/language-plutus-core.cabal:322]: I *think* the only difference between an executable and a benchmark is that a benchmark gets run when you do `cabal bench` (seems you can `cabal run` benchmarks, so that's fine). I'm not sure that we want that, so I think it's fine for it to be an executable. Nobody is going to get confused, I don't think.

File: plutus-ledger/src/Ledger/Scripts.hs (hunk 15 above)
Comment (reactormonk) [reply plutus-ledger/src/Ledger/Scripts.hs:61]: I copy/paste the nix expression from a CONTRIBUTING IIRC

File: language-plutus-core/language-plutus-core.cabal (hunk 10 above)
[the remark this answers is not in the document]
Comment (reactormonk) [reply language-plutus-core/language-plutus-core.cabal:372]: Tested it, benchmark works nicely. Doesn't get run on `test`, and it gets the PWD right.

File: language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExBudgeting.hs (hunk 4 above)
Comment (reactormonk) [reply language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExBudgeting.hs:233]: Yeap, see `CostModel.md`.

**Comment by michaelpj** [comment]
Need to update the nix files.
